main.cc

```diff
@@ -1194,6 +1194,10 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             replica::distributed_loader::init_system_keyspace(sys_ks, db, ss, gossiper, raft_gr, *cfg, system_table_load_phase::phase2).get();
 
             if (raft_gr.local().is_enabled()) {
+                if (!db.local().uses_schema_commitlog()) {
```

Comment:
I think it is reasonable to ask to shutdown node cleanly during rolling upgrade.

Reply:
> I think it is reasonable to ask to shutdown node cleanly during rolling upgrade.

We need to detect somehow that the current boot is a first boot after moving raft tables to schema commitlog. As @tgrabiec mentioned above, we need to either look into the contents of commitlog while replaying it, or remember somewhere that the raft tables have already moved to a schema commitlog.

Reply:
@tgrabiec for schema tables I also see [some logic with truncation records](https://github.com/scylladb/scylladb/blob/master/main.cc#L1236). Do we need something similar for raft tables?

Reply:
> > I think it is reasonable to ask to shutdown node cleanly during rolling upgrade.
> 
> We need to detect somehow that the current boot is a first boot after moving raft tables to schema commitlog. As @tgrabiec mentioned above, we need to either look into the contents of commitlog while replaying it, or remember somewhere that the raft tables have already moved to a schema commitlog.

I am saying that we may assume a clean shutdown during upgrade, so there cannot be a situation where non replayed raft data is in the old commitlog. Of course detecting such case would be a bonus, but I am not convinced the it is worth the complication. Currently it is not a problem even if it happens, and if it will not become a problem before 5.3 release it will never will since any upgrade will go through 5.3.

Reply:
> @tgrabiec for schema tables I also see [some logic with truncation records](https://github.com/scylladb/scylladb/blob/master/main.cc#L1236). Do we need something similar for raft tables?

Yes, if we want to be extra safe.

Reply:
> Yes, if we want to be extra safe.

I think I agree with Gleb, we can describe the procedure for upgrading to 5.3 in the documentation and assume there will be no commit log replay during node boot.

Reply:
@tgrabiec Considering Gleb’s arguments, do you agree to ask user for a clean rolling restart in the docs without enforcing this in the code? This could also make https://github.com/scylladb/scylladb/pull/13263 simpler.

In any case, it’s worth mentioning somewhere in the documentation for the 5.3 upgrade procedure that we need a clean rolling restart because of this feature. Could you advise where/how to do this? Maybe some example.

Reply:
We already ask for a clean shutdown in the docs for upgrades. 

Small correction: even if users shutdown cleanly, you can assume that all data is in the sstables, but not that there is no commitlog replay. The segments may be still kept alive for some reason. We shouldn't assume no replay, but I don't think this is problematic if all writes are already in the sstables. Commitlog replay doesn't annotate mutations with replay positions currently, it passes an empty `db::rp_handle`.

I'm not sure it's good to trust users in that regard. One thing is to require something, another is assuming without checking that users do something. Users are notoriously unreliable, so it's good to be defensive. @avikivity WDYT?

Reply:
Are we guaranteed to flush the memtables and not replay commitlog after a clean shutdown?

IMO we should stay on the safe side if there's a reasonable not-too-hard way to do that. Don't trust the user they won't f* up.

> Could you advise where/how to do this? Maybe some example.

The documentation for 5.2->5.3 upgrade does not exist yet, we will probably create it after we release 5.2.
Once it's created, it will be similar to this one (this is for 5.1->5.2): https://github.com/scylladb/scylladb/blob/master/docs/upgrade/upgrade-opensource/upgrade-guide-from-5.1-to-5.2/upgrade-guide-from-5.1-to-5.2-generic.rst
and we'll be able to update it in a regular pull request.

We could also update the 5.1->5.2 guide to mention the necessity for clean shutdown somewhere. The 5.2->5.3 guide will most probably be created initially by copying the 5.1->5.2 guide.

Reply:
the simplest way to implement the check (IIUC):
* add a bool flag `raft_tables_uses_schema_commitlog` to some schema commitlog-managed table, e.g. as a static column in `system.raft` table with false as a default value.
* check this flag in `main.cc` after replaying schema commitlog
* if the flag is true, do nothing
* if the flag is false, check if there is something to replay in the regular commitlog (`!db.local().commitlog()->get_segments_to_replay().empty()`), fail the boot with exception if it is. User will need to rollback to previous version and ensure clean restart.
* otherwise, save true to the flag and proceed.

I'm still not sure we really need this, since right now there is no problem if we just replay the raft tables from the normal commitlog. We can add this check in a separate PR, if someone needs raft group0 before the regular commitlog replay.

Reply:
> 
> I'm still not sure we really need this, since right now there is no problem if we just replay the raft tables from the normal commitlog. We can add this check in a separate PR, if someone needs raft group0 before the regular commitlog replay.

I agree and I am do not see why the PR will be ever needed. If it is fine to replay old commitlog in 5.3 and if the dependency of regular commitlog reply on  group0 will be introduced afterwords then upgrade path is covered since it has to go through 5.3.

Reply:
If we require upgrade through 5.3 (with this PR), and we won't move group0 earlier in 5.3, then there is no problem.

How can we enforce staged upgrades like this? All upgrade guides of later versions will have to mention this. @mykaul  @annastuchlik

Reply:
I think we already require upgrade to not jump over minor versions.